tools/wptrunner/wptrunner/wptrunner.py

```diff
@@ -235,143 +360,83 @@ def run_tests(config, test_paths, product, **kwargs):
             recording.set(["startup", "ensure_environment"])
             try:
                 test_environment.ensure_started()
+                start_time = datetime.now()
             except env.TestEnvironmentError as e:
                 logger.critical("Error starting test environment: %s" % e)
                 raise
 
             recording.set(["startup"])
 
+            max_time = None
+            if "repeat_max_time" in kwargs:
+                max_time = timedelta(minutes=kwargs["repeat_max_time"])
+
             repeat = kwargs["repeat"]
-            repeat_count = 0
             repeat_until_unexpected = kwargs["repeat_until_unexpected"]
 
-            while repeat_count < repeat or repeat_until_unexpected:
-                repeat_count += 1
+            # keep track of longest time taken to complete a
+            # test suite iteration so that the runs can be stopped
+            # to avoid a possible TC timeout.
+            longest_iteration_time = timedelta()
+            # keep track if we break the loop to avoid timeout.
+            iteration_timeout = False
+
+            while counts["repeat"] < repeat or repeat_until_unexpected:
+                # if the next repeat run could cause the TC timeout to be
+                # reached, stop now and use the test results we have.
+                estimate = datetime.now() + longest_iteration_time
+                if not repeat_until_unexpected and max_time \
+                        and estimate >= start_time + max_time:
+                    iteration_timeout = True
+                    logger.info(
+                        "Repeat runs are in danger of reaching timeout!"
+                        " Quitting early.")
+                    logger.info(
+                        "Ran %s of %s iterations." %
+                        (counts["repeat"], repeat))
+                    break
+
+                # begin tracking runtime of the test suite
+                iteration_start = datetime.now()
+                counts["repeat"] += 1
                 if repeat_until_unexpected:
-                    logger.info("Repetition %i" % (repeat_count))
+                    logger.info("Repetition %i" % (counts["repeat"]))
                 elif repeat > 1:
-                    logger.info("Repetition %i / %i" % (repeat_count, repeat))
-
-                test_count = 0
-                unexpected_count = 0
-                unexpected_pass_count = 0
-
-                tests = []
-                for test_type in test_loader.test_types:
-                    tests.extend(test_loader.tests[test_type])
-
-                try:
-                    test_groups = test_source_cls.tests_by_group(tests, **test_source_kwargs)
-                except Exception:
-                    logger.critical("Loading tests failed")
+                    logger.info(
```

Comment:
I don't really know why this ended up getting reformatted, by my count the original line was 78 columns (I think we enforce 100 columns in pyflakes, but certainly more than this).

Reply:
Formatting fixed